### PR TITLE
[SPARK-35176][PYTHON] Standardize input validation error type

### PR DESCRIPTION
**Pull request by Yikun**
### What changes were proposed in this pull request?
This PR corrects some exception type when the function input params are failed to validate due to TypeError.
In order to convenient to review, there are 3 commits in this PR:
- Standardize input validation error type on sql
- Standardize input validation error type on ml
- Standardize input validation error type on pandas

### Why are the changes needed?
As suggestion from Python exception doc [1]: "Raised when an operation or function is applied to an object of inappropriate type.", but there are many Value error are raised in some pyspark code, this patch fix them.

[1] https://docs.python.org/3/library/exceptions.html#TypeError

Note that: this patch only addresses the exsiting some wrong raise type for input validation, the input validation decorator/framework which mentioned in [SPARK-35176](https://issues.apache.org/jira/browse/SPARK-35176), would be submited in a speparated patch.

### Does this PR introduce _any_ user-facing change?
Yes, code can raise the right TypeError instead of ValueError.

### How was this patch tested?
Existing test case and UT

### CODE REVIEW

**Comment by Yikun** [comment]
cc @HyukjinKwon

**Comment by HyukjinKwon** [comment]
Nice, thanks for working on this @Yikun. I am off this week so might be less active .. 
cc @zero323 and @viirya FYI

**Comment by zero323** [comment]
At first glance it looks good (I'll try to do more thorough scan if I have access to larger screen). We might have to document this as a change of behaviour though, as it might break some 3rd party code.

**Comment by HyukjinKwon** [comment]
Oh yeah, that's a very good point. 

@Yikun feel free to create a new page for Spark 3.1 -> 3.2 at https://github.com/apache/spark/tree/master/python/docs/source/migration_guide and document this one

**Comment by Yikun** [comment]
@HyukjinKwon @zero323 Thanks for reminder, the migration doc has been added in https://github.com/apache/spark/pull/32368/commits/02080b00f7641a4b71566517e76d75759d6bc3f8

File: python/docs/source/migration_guide/pyspark_3.1_to_3.2.rst [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    software distributed under the License is distributed on an
+    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+    KIND, either express or implied.  See the License for the
+    specific language governing permissions and limitations
+    under the License.
+
+
+=================================
+Upgrading from PySpark 3.1 to 3.2
+=================================
```
Comment (viirya) [inline python/docs/source/migration_guide/pyspark_3.1_to_3.2.rst:21]: Don't we need to put this to `index.rst`?

**Review by viirya — COMMENTED** [review]
Looks okay as TypeError should be more accurate for the kind of exceptions.

File: python/docs/source/migration_guide/pyspark_3.1_to_3.2.rst (hunk 1 above)
Comment (Yikun) [reply python/docs/source/migration_guide/pyspark_3.1_to_3.2.rst:21]: done

**Review by HyukjinKwon — APPROVED** [review]

**Review by viirya — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**